File: DSA/to_revise/1749 - maxAbsoluteSum.cpp
// https://leetcode.com/problems/maximum-absolute-sum-of-any-subarray/

#include <bits/stdc++.h>
using namespace std;

class Solution {
    public:
        int maxAbsoluteSum(vector<int>& nums) {
            int res(0), maxPref(0), minPref(0), curSum(0);
            for (int n : nums) {
                curSum += n;
                res = max(res, max(abs(curSum - maxPref), abs(curSum - minPref)));
                maxPref = max(maxPref, curSum);
                minPref = min(minPref, curSum);
            }   
            return res;
        }
};

File: DSA/to_revise/1749 - maxAbsoluteSum.py
# https://leetcode.com/problems/maximum-absolute-sum-of-any-subarray/

from typing import List

class Solution:
    def maxAbsoluteSum(self, nums: List[int]) -> int:
        min_pre = res = max_pre = cur = 0
        for n in nums:
            cur += n
            res = max(res, abs(cur - min_pre), abs(cur - max_pre))
            min_pre = min(min_pre, cur)
            max_pre = max(max_pre, cur)
        return res
